Database set up

souravdhar199 opened the pull request:
I just want to make sure that I'm doing the right thing. I tried to define some relationship between some tables, please review my models.py file and let me know If I did anything wrong

souravdhar199 committed:
feat: Created a models.py file, also created some tables and tried to do some relationship betweens some attribute

omarkmu commented on app.py:
```diff
@@ -72,7 +76,7 @@ def login():
-        host=os.getenv("HOST", "0.0.0.0"),
-        port=int(os.getenv("PORT", 8080)),
+        # host=os.getenv("HOST", "0.0.0.0"),
+        # port=int(os.getenv("PORT", 8080)),
```
Why were these lines commented out? You should try adding the necessary environment variables to your `.env` if you're having trouble running it.

omarkmu commented on models.py:
```diff
@@ -0,0 +1,45 @@
+from app import db
+
+
+class owner(db.Model):
```
This should be PascalCase `Owner`, not lowercase. I think a name like `BTUser` (BotTrition User) would be clearer, if we're avoiding using `User`.

omarkmu commented on models.py:
```diff
@@ -0,0 +1,45 @@
+class owner(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    username = db.Column(db.String(80))
+    password = db.Column(db.String(80))
```
It might be better to change the name of this field to make it clear that it's a *hashed* password (`password_hash`?).

omarkmu commented on models.py:
```diff
@@ -0,0 +1,45 @@
+    foodAllergies = db.relationship("Allergies", backref="food_allergies")
+    dietaryrestriction = db.relationship("DietaryRestriction", backref="food_dietary")
+    ownerage = db.relationship("Age", backref="user_age", uselist=False)
+    profileData = db.relationship("Age", backref="user_data", uselist=False)
```
Good job setting these up! A few problems, though:

- The case on each of these fields should be `snake_case`. For example, `foodAllergies` should be `food_allergies`.
- You seem to have used `"Age"` as the table name a second time on line 12. Did you mean `"Profile"`?
- The one-to-one relationships here don't really need to be in a separate table. The fields in `Age` and `Profile` could just be columns in the User table.

omarkmu commented on models.py:
```diff
@@ -0,0 +1,45 @@
+    day = db.Column(db.Integer)
+    month = db.Column(db.Integer)
+    year = db.Column(db.Integer)
+    user_calculated_age = db.Column(db.Integer, db.ForeignKey("user.id"))
```
We can calculate this as needed in the code. I don't think we need to include a field for this.

omarkmu commented on models.py:
```diff
@@ -0,0 +1,45 @@
+
+class Allergies(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    users_food_allergies = db.Column(db.Integer, db.ForeignKey("owner.id"))
```
There doesn't seem to be a field for the allergy itself in this table. There's a primary key and foreign key, but no field is included to track the allergy. I recommend renaming this field to `user_id` and adding an `allergy` field.

omarkmu commented on models.py:
```diff
@@ -0,0 +1,45 @@
+
+class DietaryRestriction(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    users_food_dietary = db.Column(db.Integer, db.ForeignKey("owner.id"))
```
Similar suggestion as the one on line 29: The restriction itself is missing from this table. I think a good name for a new field might be `restriction` or `food`. I would also rename this field to `user_id` to be clearer.

souravdhar199 committed:
fix: Fixed some issues with renaming variable and few bugs.

omarkmu committed:
fix: Modified database fields and relationships

Made slight corrections to relationships, field names, and table names.
Also, merged the Age and Profile tables into one table.

omarkmu committed:
refactor: Renamed models.py to database.py

Gave models.py a broader name in preparation for structure modification

omarkmu committed:
fix: Extracted database initialization
refactor: Renamed user_profile field in BTUser to profile

Database initialization is now handled in `database.py`.
This should help to avoid circular import errors.

omarkmu committed:
refactor: Merge allergies table into restrictions
refactor: Rename dietary_restrictions field to restrictions

BrandonDwyer commented on database.py:
```diff
@@ -0,0 +1,53 @@
+    user_id = db.Column(db.Integer, db.ForeignKey("btuser.id"))
+
+    gender = db.Column(db.String(80))
+    height = db.Column(db.Integer)
```
I see why this is an integer to have an easier time calculating for the BMI. However, what about the possibility for this to be a string rather than an integer, because a user maybe 6 feet tall but might not know the conversion to add 72inches in the table to put as their height. Some issues of switching to a string will be having to get the height from the string.

omarkmu replied:
I think we should perform that conversion on the user's input. The frontend could offer two separate fields for the feet and inches components of height. For example, two textboxes separated by `'` and suffixed by `"` with numbers as the placeholder text could signal to the user that feet and inches are expected. Then, we could convert the height to a single integer before storing it.

BrandonDwyer commented on database.py:
```diff
@@ -0,0 +1,53 @@
+    weight = db.Column(db.Integer)
+
+    birth_day = db.Column(db.Integer)
+    birth_month = db.Column(db.Integer)
```
This can be considered a string value depending on how to design the user data entry for their individual data.

BrandonDwyer commented on database.py:
```diff
@@ -0,0 +1,53 @@
+    id = db.Column(db.Integer, primary_key=True)
+    username = db.Column(db.String(80), unique=True, nullable=False)
+    password_hash = db.Column(db.String(80), nullable=False)
+
```
We could add more specificity to the restriction if needed, or we could do this later. Maybe add religious_restriction or diet restrictions like keto, vegan, or pescatarian. However, though this sprint covers the MVP, these extra fields are unnecessary and are up to a design decision and can be implemented during the polish sprint.

BrandonDwyer commented on database.py:
```diff
@@ -0,0 +1,53 @@
+
+    id = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(db.Integer, db.ForeignKey("btuser.id"))
+    restriction_type = db.Column(db.String(10))  # one of "allergy", "diet"
+    restriction = db.Column(db.String(80))
```
As stated above in the first comment, we could add another relationship to break down diets for more modularity but what is present is fine for getting the MVP ready and working.

BrandonDwyer commented on app.py:
```diff
@@ -1,11 +1,15 @@
 from flask_login import login_user, current_user, LoginManager
 from flask_login.utils import login_required
-from flask_sqlalchemy import SQLAlchemy
+from database import db
```
Good job importing from another file to prevent possible circular imports in the future!

BrandonDwyer commented on app.py:
```diff
@@ -43,11 +47,10 @@ def index():
-#NOTE user cannot be verified until DB is set up and conditional statements 
-#will have to be updated
+# registration page to register a new user
+# NOTE user cannot be verified until DB is set up and conditional statements
+# will have to be updated
```
These comments can probably be deleted on the next pushes made to repo.

BrandonDwyer reviewed (commented):
Overall good changes to finally get the DB functionality going. See the comments for some design points and directions to possibly take for designing this app, but what is present in these changes is perfectly fine to be ready to be merged.

omarkmu committed:
Merge branch 'main' into feat/database-model
